crux_core/src/capabilities/compose.rs

```diff
@@ -0,0 +1,101 @@
+//! A capability which can spawn tasks which orchestrate across other capabilities. This
+//! is useful for orchestrating a number of different effects into a single transaction.
+
+use crate::capability::{CapabilityContext, Never};
+use crate::Capability;
+use futures::Future;
+
+/// Compose capability can be used to orchestrate effects into a single transaction.
+///
+/// Example include:
+/// * Running a number of HTTP requests in parallel and waiting for all to finish
+/// * Chaining effects together, where the output of one is the input of the next and the intermediate
+///   results are not useful to the app
+/// * Implementing request timeouts by selecting across a HTTP effect and a time effect
+/// * Any arbitrary graph of effects which depend on each other (or not).
+///
+/// Note that testing composed effects is more difficult, because it is not possible to enter the effect
+/// transaction "in the middle" - only from the beginning - or to ignore some of the effects with out
+/// stalling the entire downstream dependency chain.
+pub struct Compose<Ev> {
+    context: CapabilityContext<Never, Ev>,
+}
+
+/// A restricted context given to the closure passed to [`Compose::spawn`]. This context can only
+/// update the app, not request from the shell or spawn further tasks.
+pub struct ComposeContext<Ev> {
+    context: CapabilityContext<Never, Ev>,
+}
+
+impl<Ev> ComposeContext<Ev> {
+    /// Update the app with an event. This forwards to [`CapabilityContext::update_app`].
+    pub fn update_app(&self, event: Ev)
+    where
+        Ev: 'static,
+    {
+        self.context.update_app(event);
+    }
+}
+
+impl<Ev> Compose<Ev> {
+    pub fn new(context: CapabilityContext<Never, Ev>) -> Self {
+        Self { context }
+    }
+
+    /// Spawn a task which orchestrates across other capabilities.
+    ///
+    /// The argument is a closure which receives a [`ComposeContext`] which can be used to send
+    /// events to the app.
+    ///
+    /// For example:
+    /// ```
+    /// fn update(&self, event: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
+    ///     match event {
+    ///        Event::Trigger => caps.compose.spawn(|context| {
+    ///             let caps = Clone::clone(caps);
```

Comment:
Not a big deal, but maybe: 

```suggestion
    ///             let caps = caps.clone();
```

Don't think I've ever seen someone do `Clone::clone`, looks a bit odd

Reply:
> @StuartHarris pointed out that Clone::clone is somewhat more explicit of the intent because it doesn't rely on automatic dereferencing.

True, though I'd usually express that as `WhateverTheActualTypeIsCalled::clone`.  Although maybe that's just me.

But whatever, does not matter.

Reply:
Ah yes. I might still change it to `.clone()` because it's the least likely to trip the reader up and make them wonder what is going on.